### DISCUSSION TITLE
Migration guide - 3.x.x -> 4.x.x

### DISCUSSION THREAD

**Discussion by andyl-ioki** [fnando/i18n#24] [Q&A]
Hej there!

At first: Thanks for keeping this project going! 👍 👍 🚀 

Situation:
We updated  i18n-js from `3.9.2` to `4.0.2` and tried to look up for changes or alternatively a migration guide. But I did not found any.

Question:
Is there any?
If not, maybe I can add some sentences to the `readme` as pull request? Wdyt?

**Comment by BarbecueSilver** [comment]
So this is what I got so far and what I had to do, to make it run again 👍 

# Migration Guide 3.x.x -> 4.x.x
## Usage
### Setting up
You have to instantiate an `I18n` object now. You pass the settings along with it.
See the [documentation](https://github.com/fnando/i18n#setting-up) for more information about setting i18n-js up in version 4.x.x.
#### Before
```TypeScript
import I18n from 'i18n-js';
import english from './en.json';
import yoruba from './yo.json';
...
I18n.locale = 'en';
I18n.fallbacks = true;
I18n.translations = {
  en: english,
  yo: yoruba, 
};
```
#### After
```TypeScript
import I18n from 'i18n-js';
import english from './en.json';
import yoruba from './yo.json';
...
const i18n = new I18n(
  // Translations
  {
    en: english,
    yo: yoruba,
  },
  {
    locale: "en",
    enableFallback: true,
  }
);
```
### Translating messages
You have to use the instantiated object now.
#### Before
```TypeScript
import I18n from 'i18n-js';
...
I18n.t('main.welcome');
```

#### After
```TypeScript
import I18n from 'i18n-js';
...
export const i18n = new I18n(...);
```
```TypeScript
import { i18n } from '.../i18n.ts';
...
i18n.t('main.welcome');
```


### toCurrency
Renamed to `numberToCurrency`.

#### Before
```TypeScript
I18n.toCurrency(amount / 100, options);
```

#### After 
```TypeScript
// i18n is the initialized object, see "Setting up"
i18n.numberToCurrency(amount / 100, options);
```

## Types
### ToCurrencyOptions
Changed to `NumberToCurrencyOptions`.
Can't be exported anymore from `i18n-js`.

#### Before
```TypeScript
import { ToCurrencyOptions } from 'i18n-js';
...
const options: ToCurrencyOptions = {
  delimiter: '.',
  separator: ',',
  precision: 4,
  format: '%n %u',
};
```

#### After
```TypeScript
import { NumberToCurrencyOptions } from 'i18n-js/typings/typing';
...
const options: NumberToCurrencyOptions = {
  delimiter: '.',
  separator: ',',
  precision: 4,
  format: '%n %u',

  negativeFormat: '-%n %u',
  roundMode: 'default',
  significant: false,
  stripInsignificantZeros: false,
  raise: false,
  unit: '',
};
```

## Testing
### Mocking with jest
Check out [jest documentation](https://jestjs.io/docs/manual-mocks#mocking-node-modules) about mocking node modules.
#### Before
`Tests/__mocks__/i18n-js.ts`
```TypeScript
export default {
  locale: 'en',
  t: (translation) => translation,
  numberToCurrency: jest.fn(),
  currentLocale: () => 'yo',
}
```
#### After
`Tests/__mocks__/i18n-js.ts`
```TypeScript
export const I18n = jest.fn().mockImplementation(() => {
  return {
  locale: 'yo',
  t: (translation) => translation,
  numberToCurrency: jest.fn(),
  };
});

```

**Comment by fnando** [comment]
I added a slim migration guide to the main repo as well.

https://github.com/fnando/i18n-js/blob/main/MIGRATING_FROM_V3_TO_V4.md
